pkg/pillar/hypervisor/kvm.go

```diff
@@ -942,40 +1068,295 @@ func getFmlCustomResolution(status *types.DomainStatus, globalConfig *types.Conf
+		for j := range a.virtualNetworks {
+			if i == j {
+				continue
+			}
+			if a.virtualNetworks[j].VifOrder < a.virtualNetworks[i].VifOrder {
+				pciDeviceID++
+			}
+		}
+		for pciAddr, md := range a.multifunctionDevices {
+			isBefore, isAfter := md.compareOrderWithVirtNet(a.virtualNetworks[i])
+			invalidOrder := isBefore && isAfter
+			if invalidOrder {
+				return logError("user-defined network interface order "+
+					"for VIF %s breaks order of the multifunction PCI device %s",
+					a.virtualNetworks[i].Vif, pciAddr)
+			}
+			if isBefore {
+				pciDeviceID++
+			}
+		}
+		a.virtualNetworks[i].pciDeviceID = pciDeviceID
+	}
+
+	// Determine PCI addresses for direct PCI assignments.
+	for i := range a.pciAssignments {
+		pciLongWoFunc, err := a.pciAssignments[i].pciLongWOFunction()
+		if err != nil {
+			logrus.Warnf("retrieving pci address without function failed: %v", err)
+			continue
+		}
+		md := a.multifunctionDevices[pciLongWoFunc]
+		if md == nil {
+			// Even when device is not multifunction, it still should have entry
+			// in the multifunctionDevices map.
+			logrus.Warnf("missing multifunctionDevices entry for pci address: %s",
+				pciLongWoFunc)
+			continue
+		}
+		// Increment pciDeviceOrBridgeID by 1 for every (virtual or assigned) PCI device
+		// which should have lower PCI address.
+		// For a multifunction PCI device, we increment by 1 for the entire group of
+		// functions, as they share a single address on the root bus through their bridge.
+		pciDeviceOrBridgeID := a.firstFreePCIID
+		for _, virtNet := range a.virtualNetworks {
+			// Order validity already checked when addresses for virtual networks
+			// were determined.
+			if isBefore, _ := md.compareOrderWithVirtNet(virtNet); !isBefore {
+				// Also increased when order is undefined, i.e. this PCI device
+				// does not have network function. Non-networking PCI devices
+				// are placed after virtual network interfaces.
+				pciDeviceOrBridgeID++
+			}
+		}
+		thisHasNetFunc := md.hasNetworkFunction()
+		for pciAddr2, md2 := range a.multifunctionDevices {
+			if pciLongWoFunc == pciAddr2 {
+				continue
+			}
+			theOtherHasNetFunc := md2.hasNetworkFunction()
+			if !thisHasNetFunc {
+				if theOtherHasNetFunc {
+					// Network functions take priority in the order.
+					pciDeviceOrBridgeID++
+				} else {
+					// Between non-networking devices, order by PCI addresses
+					// lexicographically.
+					if pciLongWoFunc > pciAddr2 {
+						pciDeviceOrBridgeID++
+					}
+				}
+				continue
+			}
+			if !theOtherHasNetFunc {
+				// The other non-network device is ordered after this network device.
+				continue
+			}
+			// Both devices have at least one network function.
+			theOtherIsBefore, theOtherIsAfter := md2.compareOrder(*md)
+			invalidOrder := theOtherIsBefore && theOtherIsAfter
+			if invalidOrder {
+				return logError("user-defined network interface order "+
```

Comment:
We should remember that this message will appear in the UI and be seen by the user. Does it explain in simple words what the user did wrong? 
To be honest, I find it hard to describe in short words the limitations that we face here. I think it would be perfect to link a longer explanation of the restriction here, but I have no idea where we can point to.

Reply:
How about this:

return logError("Invalid VIF %s configuration: user-defined network interface order disrupts the function sequence of the multifunction PCI device %s. Interleaving PCI device functions with other devices is not allowed.", a.virtualNetworks[i].Vif, pciAddr)

Reply:
I'm going to go with this new error message - but I'm also open to new suggestions.